mlir/tools/mlir-clang/Lib/clang-mlir.cc

```diff
@@ -565,26 +537,60 @@ ValueWithOffsets MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
     }
   if (auto ic = dyn_cast<ImplicitCastExpr>(expr->getCallee()))
     if (auto sr = dyn_cast<DeclRefExpr>(ic->getSubExpr())) {
-      // TODO add pow to standard dialect
-      if (sr->getDecl()->getName() == "fprintf") {
-        llvm::errs() << "warning skipping fprintf\n";
-        // std::vector<mlir::Value> args;
-        // for(auto a : expr->arguments()) {
-        //    args.push_back((mlir::Value)Visit(a));
-        //}
-        return nullptr;
-      }
-      if (sr->getDecl()->getName() == "printf") {
-        llvm::errs() << "warning skipping printf\n";
-        // std::vector<mlir::Value> args;
-        // for(auto a : expr->arguments()) {
-        //    args.push_back((mlir::Value)Visit(a));
-        //}
+      if (sr->getDecl()->getName() == "fprintf" ||
+          sr->getDecl()->getName() == "printf") {
+
+        auto tocall = EmitCallee(expr->getCallee());
+        auto fprintfF = Glob.GetOrCreateLLVMFunction(tocall);
+
+        std::vector<mlir::Value> args;
+        size_t i = 0;
+        for (auto a : expr->arguments()) {
+          if (i == 0 && sr->getDecl()->getName() == "fprintf") {
+            auto decl =
+                cast<DeclRefExpr>(cast<ImplicitCastExpr>(a)->getSubExpr());
+            args.push_back(builder.create<mlir::LLVM::AddressOfOp>(
+                loc,
+                Glob.GetOrCreateLLVMGlobal(cast<VarDecl>(decl->getDecl()))));
+            i++;
+            continue;
+          }
+
+          if (auto IC1 = dyn_cast<ImplicitCastExpr>(a)) {
+            if (auto IC2 = dyn_cast<ImplicitCastExpr>(IC1->getSubExpr())) {
+              if (auto slit =
+                      dyn_cast<clang::StringLiteral>(IC2->getSubExpr())) {
+                args.push_back(Glob.GetOrCreateGlobalLLVMString(
+                    loc, builder, slit->getString()));
+                i++;
+                continue;
+              }
+            }
+          }
+          if (auto IC1 = dyn_cast<ImplicitCastExpr>(a)) {
```

Comment:
This check is the same as the one above (line 559). Can you not move lines 571 to 575 here, and remove the duplicate check?

Reply:
Done